src/USER-CONP/fix_charge_update.cpp

```diff
@@ -0,0 +1,402 @@
+      std::vector<std::vector<double>> mat(ngroup, std::vector<double>(ngroup));
+      for (bigint i = 0; i < ngroup; i++) {
+        bigint const gi = group_idx[i];
+        for (bigint j = 0; j < ngroup; j++) {
+          mat[gi][group_idx[j]] = a[i][j];
+        }
+      }
+      write_to_file(f_ela, taglist_bygroup, mat);
+    }
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixChargeUpdate::create_taglist() {
+  int *mask = atom->mask;
+  int const nlocal = atom->nlocal;
+  int const nprocs = comm->nprocs;
+  tagint *tag = atom->tag;
+
+  // assign a tag to each matrix index sorted by group and by tag
+  taglist_bygroup = std::vector<tagint>();
+  for (int gbit : group_bits) {
+    std::vector<tagint> taglist_local;
+    for (int i = 0; i < nlocal; i++) {
+      if (mask[i] & gbit) taglist_local.push_back(tag[i]);
+    }
+    // gather from all cpus for this group
+    int gnum_local = taglist_local.size();
+    std::vector<int> idispls(nprocs);
+    std::vector<int> gnum_list(nprocs);
+    MPI_Allgather(&gnum_local, 1, MPI_INT, &gnum_list.front(), 1, MPI_INT,
+                  world);
+    idispls[0] = 0;
+    for (int i = 1; i < nprocs; i++) {
+      idispls[i] = idispls[i - 1] + gnum_list[i - 1];
+    }
+    int const gnum = accumulate(gnum_list.begin(), gnum_list.end(), 0);
+    std::vector<tagint> taglist_all(gnum);
+    MPI_Allgatherv(&taglist_local.front(), gnum_local, MPI_LMP_TAGINT,
+                   &taglist_all.front(), &gnum_list.front(), &idispls.front(),
+                   MPI_LMP_TAGINT, world);
+    std::sort(taglist_all.begin(), taglist_all.end());
+    for (tagint t : taglist_all) taglist_bygroup.push_back(t);
+  }
+
+  // taglist only sorted by tag not group, same order as in computes
+  taglist = taglist_bygroup;
+  std::sort(taglist.begin(), taglist.end());
+
+  // group_idx allows mapping a vector that is sorted by taglist to being
+  // ordered by taglist_bygroup
+  group_idx = std::vector<tagint>();
+  for (tagint t : taglist) {
+    for (size_t i = 0; i < taglist_bygroup.size(); i++) {
+      if (t == taglist_bygroup[i]) {
+        group_idx.push_back(i);
+        break;
+      }
+    }
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+std::vector<int> FixChargeUpdate::local_to_matrix() {
+  int const nall = atom->nlocal + atom->nghost;
+  int *tag = atom->tag;
+  std::vector<int> mpos(nall, -1);
+  for (bigint ii = 0; ii < ngroup; ii++) {
+    for (int i = 0; i < nall; i++) {
+      if (taglist[ii] == tag[i]) {
+        mpos[i] = ii;
+      }
+    }
+  }
+  return mpos;
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixChargeUpdate::pre_force(int) {
+  std::vector<int> mpos = local_to_matrix();
+  int const nall = atom->nlocal + atom->nghost;
+  double **a = read_matrix ? matrix_from_file : array_compute->array;
```

Comment:
Das verstehe ich noch nicht ganz. Ist das der Grund warum du array in compute_conp_matrix.cpp brauchst?

Reply:
Ja, dadurch dass array zu compute.h gehört, kann ich vom fix aus darauf zugreifen.

Reply:
Ah OK! Kannst du dann nicht gradQ_V und array zusammenfassen? Eigentlich brauchen wir ja keine zwei Matrizen oder?

Reply:
array ist aber die invertierte gradQ_V. Ich habe auch schon überlegt, die nicht invertierte Matri an den Fix zu übergeben, dann hätte man die Möglichkeit die uninvertierte Matrix auszugeben

Reply:
Ich würde das begrüßen. Wäre etwas sauberer. Denn nur der fix braucht ja eigentlich die invertierte Matrix, deshalb solte die eigentlich auch erst dort invertiert werden.

Reply:
Ja, ich pass das an